kbfsblock/protocol_utils.go

```diff
@@ -0,0 +1,229 @@
+	// Block references to MD blocks are allowed, because they can be
+	// deleted in the case of an MD put failing.
+	return keybase1.BlockReference{
+		Bid: makeIDCombo(id, context),
+		// The actual writer to modify quota for.
+		ChargedTo: context.GetWriter(),
+		Nonce:     keybase1.BlockRefNonce(context.GetRefNonce()),
+	}
+}
+
+// MakeGetBlockArg builds a keybase1.GetBlockArg from the given params.
+func MakeGetBlockArg(tlfID tlf.ID, id ID, context Context) keybase1.GetBlockArg {
+	return keybase1.GetBlockArg{
+		Bid:    makeIDCombo(id, context),
+		Folder: tlfID.String(),
+	}
+}
+
+// ParseGetBlockRes parses the given keybase1.GetBlockRes into its
+// components.
+func ParseGetBlockRes(res keybase1.GetBlockRes, resErr error) (
+	buf []byte, serverHalf kbfscrypto.BlockCryptKeyServerHalf, err error) {
+	if resErr != nil {
+		return nil, kbfscrypto.BlockCryptKeyServerHalf{}, resErr
+	}
+	serverHalf, err = kbfscrypto.ParseBlockCryptKeyServerHalf(res.BlockKey)
+	if err != nil {
+		return nil, kbfscrypto.BlockCryptKeyServerHalf{}, err
+	}
+	return res.Buf, serverHalf, nil
+}
+
+// MakePutBlockArg builds a keybase1.PutBlockArg from the given params.
+func MakePutBlockArg(tlfID tlf.ID, id ID,
+	bContext Context, buf []byte,
+	serverHalf kbfscrypto.BlockCryptKeyServerHalf) keybase1.PutBlockArg {
+	return keybase1.PutBlockArg{
+		Bid: makeIDCombo(id, bContext),
+		// BlockKey is misnamed -- it contains just the server
+		// half.
+		BlockKey: serverHalf.String(),
+		Folder:   tlfID.String(),
+		Buf:      buf,
+	}
+}
+
+// MakePutBlockAgainArg builds a keybase1.PutBlockAgainArg from the
+// given params.
+func MakePutBlockAgainArg(tlfID tlf.ID, id ID,
+	bContext Context, buf []byte, serverHalf kbfscrypto.BlockCryptKeyServerHalf) keybase1.PutBlockAgainArg {
+	return keybase1.PutBlockAgainArg{
+		Ref: makeReference(id, bContext),
+		// BlockKey is misnamed -- it contains just the server
+		// half.
+		BlockKey: serverHalf.String(),
+		Folder:   tlfID.String(),
+		Buf:      buf,
+	}
+}
+
+// MakeAddReferenceArg builds a keybase1.AddReferenceArg from the
+// given params.
+func MakeAddReferenceArg(tlfID tlf.ID, id ID, context Context) keybase1.AddReferenceArg {
+	return keybase1.AddReferenceArg{
+		Ref:    makeReference(id, context),
+		Folder: tlfID.String(),
+	}
+}
+
+// getNotDone returns the set of block references in "all" that do not
+// yet appear in "results"
+func getNotDone(all ContextMap, doneRefs map[ID]map[RefNonce]int) (
+	notDone []keybase1.BlockReference) {
+	for id, idContexts := range all {
+		for _, context := range idContexts {
+			if _, ok := doneRefs[id]; ok {
+				if _, ok1 := doneRefs[id][context.GetRefNonce()]; ok1 {
+					continue
+				}
+			}
+			ref := makeReference(id, context)
+			notDone = append(notDone, ref)
+		}
+	}
+	return notDone
+}
+
+// BatchDowngradeReferences archives or deletes a batch of references,
+// handling all batching and throttles.
+func BatchDowngradeReferences(ctx context.Context, log logger.Logger,
```

Comment:
It's not great that this BlockServer logic is here, but it's better than having to duplicate this logic on the server side.